arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -68,73 +73,218 @@ void ClusterMetricsFeature::validateOptions(
-      [this](futures::Try<RawDBServers>&& metrics) mutable {
-        // We want more time than kTimeout should expire
-        // after last try cross cluster communication
-        _lastUpdate = std::chrono::steady_clock::now();
-        if (!metrics.hasValue()) {
-          scheduleUpdate();
+      [this, version, &ci](futures::Try<RawDBServers>&& raw) mutable {
+        if (wasStop()) {
           return;
         }
+        auto result = Result::Old;
         try {
-          set(std::move(*metrics));
+          _lastUpdate = std::chrono::steady_clock::now();
+          result = writeData(version, std::move(raw));
+          _lastUpdate = std::chrono::steady_clock::now();
         } catch (...) {
+          result = Result::Error;
         }
-        repeatUpdate();
+        _acquiredUpdateLock = !ci.unlockMetricsUpdate();
+        _lastUpdate = std::chrono::steady_clock::now();
+        if (result != Result::Error) {
+          repeatUpdate();
+        }
+        rescheduleUpdate(std::max(_timeout, uint32_t{1}));
       });
 }
 
+ClusterMetricsFeature::Result ClusterMetricsFeature::readData(
+    uint64_t& version) {
+  auto vocbase = server().getFeature<SystemDatabaseFeature>().use();
+  aql::QueryString readText{
+      absl::StrCat("FOR m IN ", StaticStrings::MetricsCollection, " RETURN m")};
+  auto readQuery = aql::Query::create(
+      transaction::StandaloneContext::Create(*vocbase), readText, nullptr);
+  auto r = readQuery->executeSync();
+  if (r.fail()) {
+    return Result::Error;
+  }
+  auto metrics = r.data->slice().at(0);
+  version = metrics.get("version").getNumber<uint64_t>();
+  auto data = std::atomic_load_explicit(&_data, std::memory_order_acquire);
+  if ((data ? data->version : 0) < version) {
+    auto raw = metrics.get("data");
+    std::atomic_store_explicit(&_data, Data::fromVPack(version, raw),
+                               std::memory_order_release);
+    return Result::New;
+  }
+  return Result::Old;
+}
+
+ClusterMetricsFeature::Result ClusterMetricsFeature::writeData(
+    uint64_t version, futures::Try<RawDBServers>&& raw) {
+  auto vocbase = server().getFeature<SystemDatabaseFeature>().use();
+  if (!raw.hasValue()) {
+    return Result::Error;
+  }
+  auto metrics = parse(std::move(*raw));
+  if (metrics.values.empty()) {
+    return Result::Error;
+  }
+  VPackBuilder builder;
+  builder.openObject();
+  builder.add("d", VPackSerialize{metrics});
+  builder.add("v", VPackValue{version + 1});
+  builder.close();
+  aql::QueryString writeText{
+      absl::StrCat("FOR m IN ", StaticStrings::MetricsCollection,
+                   " FILTER m.version < @v"
+                   " REPLACE m WITH { data: @d, version: @v } IN _metrics"
+                   " RETURN 0")};
+  std::shared_ptr<VPackBuilder> sharedDefault;
+  std::shared_ptr<VPackBuilder> bindVars{sharedDefault, &builder};
+  auto writeQuery = aql::Query::create(
+      transaction::StandaloneContext::Create(*vocbase), writeText, bindVars);
```

Comment:
same here, we can avoid heap allocation for `transaction::StandaloneContext`

Reply:
Ok